modules/atmosphere/rendering/atmospheredeferredcaster.cpp

```diff
@@ -1190,369 +1000,565 @@ void AtmosphereDeferredcaster::executeCalculations(GLuint quadCalcVAO,
+                    "the not same for all attached textures; or , if the attached images are "
+                    "a mix of renderbuffers and textures, the value of "
+                    "GL_RENDERBUFFE_r_samples does not match the value of "
+                    "GL_TEXTURE_SAMPLES.");
+                LERROR("Returned if the value of GL_TEXTURE_FIXED_SAMPLE_LOCATIONS is not "
+                    "the same for all attached textures; or , if the attached images are a "
+                    "mix of renderbuffers and textures, the value of "
+                    "GL_TEXTURE_FIXED_SAMPLE_LOCATIONS is not GL_TRUE for all attached "
+                    "textures.");
+                break;
+            case GL_FRAMEBUFFER_INCOMPLETE_LAYER_TARGETS:
+                LERROR("Returned if any framebuffer attachment is layered, and any populated "
+                    "attachment is not layered, or if all populated color attachments are "
+                    "not from textures of the same target.");
+                break;
+            default:
+                LDEBUG("No error found checking framebuffer: " + codePosition);
+                break;
+            }
         }
     }
-}
-
-void AtmosphereDeferredcaster::renderQuadForCalc(GLuint vao, GLsizei numberOfVertices) {
-    glBindVertexArray(vao);
-    glDrawArrays(GL_TRIANGLES, 0, numberOfVertices);
-    glBindVertexArray(0);
-}
-
-void AtmosphereDeferredcaster::step3DTexture(
-                                std::unique_ptr<ghoul::opengl::ProgramObject>& shaderProg,
-                                int layer, bool doCalc)
-{
-    // See OpenGL redbook 8th Edition page 556 for Layered Rendering
-    if (doCalc) {
-        float earth2  = _atmospherePlanetRadius * _atmospherePlanetRadius;
-        float atm2    = _atmosphereRadius * _atmosphereRadius;
-        float diff    = atm2 - earth2;
-        float ri      = static_cast<float>(layer) / static_cast<float>(_r_samples - 1);
-        float ri_2    = ri * ri;
-        float epsilon =
-            (layer == 0) ?
-            0.01f :
-            (layer == (static_cast<int>(_r_samples) - 1)) ? -0.001f : 0.0f;
-        float r       = sqrtf(earth2 + ri_2 * diff) + epsilon;
-        float dminG   = r - _atmospherePlanetRadius;
-        float dminT   = _atmosphereRadius - r;
-        float dh      = sqrtf(r * r - earth2);
-        float dH      = dh + sqrtf(diff);
-
-        shaderProg->setUniform("r", r);
-        shaderProg->setUniform("dhdH", dminT, dH, dminG, dh);
+
+    void AtmosphereDeferredcaster::renderQuadForCalc(GLuint vao, GLsizei numberOfVertices) {
+        glBindVertexArray(vao);
+        glDrawArrays(GL_TRIANGLES, 0, numberOfVertices);
+        glBindVertexArray(0);
     }
 
-    shaderProg->setUniform("layer", static_cast<int>(layer));
-}
+    void AtmosphereDeferredcaster::step3DTexture(
+        std::unique_ptr<ghoul::opengl::ProgramObject>& shaderProg,
```

Comment:
Better to pass `ghoul::opengl::ProgramObject&`;  this function doesn't need to know how the program is stored

Reply:
I think it is clear this way. :-)